### ISSUE TITLE
Build instructions

### ISSUE THREAD

**Issue by tukusejssirs** [romcal/romcal#122]
I’ve just wanted to build `romcal` in `romcal-es6` branch, so I run `npm run build` (as it is written in the `README.md`), but it failed:

```bash
# Run in the repo root
$ npm run build

> romcal@1.3.0-beta3 build /home/ts/git/c10n/romcal
> npm run build:babel && npm run build:webpack


> romcal@1.3.0-beta3 build:babel /home/ts/git/c10n/romcal
> babel src --out-dir dist

sh: babel: command not found
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! romcal@1.3.0-beta3 build:babel: `babel src --out-dir dist`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the romcal@1.3.0-beta3 build:babel script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/ts/.npm/_logs/2019-12-14T22_48_11_542Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! romcal@1.3.0-beta3 build: `npm run build:babel && npm run build:webpack`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the romcal@1.3.0-beta3 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/ts/.npm/_logs/2019-12-14T22_48_11_559Z-debug.log
```

@emagnier, is there anything else needed to make a successful build?

**Comment by emagnier** [comment]
Just a note that it's working on travis. Did you made an `npm install`?

**Comment by tukusejssirs** [comment]
Oh, well, I have forgotten. :smiley:

But still, there are some errors and `npm` suggests to run `npm audit fix`, so I think there might be some kind of problem.

```bash
$ npm install

> core-js@3.4.8 postinstall /home/ts/git/c10n/romcal/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"

Thank you for using core-js ( https://github.com/zloirock/core-js ) for polyfilling JavaScript standard library!

The project needs your help! Please consider supporting of core-js on Open Collective or Patreon: 
> https://opencollective.com/core-js 
> https://www.patreon.com/zloirock 

Also, the author of core-js ( https://github.com/zloirock ) is looking for a good job -)


> romcal@1.3.0-beta3 install /home/ts/git/c10n/romcal
> npm run build


> romcal@1.3.0-beta3 build /home/ts/git/c10n/romcal
> npm run build:babel && npm run build:webpack


> romcal@1.3.0-beta3 build:babel /home/ts/git/c10n/romcal
> babel src --out-dir dist

Successfully compiled 79 files with Babel.

> romcal@1.3.0-beta3 build:webpack /home/ts/git/c10n/romcal
> webpack-cli --config webpack.config.js

Hash: eb3b0fd7c9386b4aab6e
Version: webpack 4.41.2
Time: 18666ms
Built at: 12/15/2019 12:08:14 AM
                   Asset      Size  Chunks                          Chunk Names
    romcal.bundle.min.js   677 KiB       0  [emitted]        [big]  romcal
romcal.bundle.min.js.map  2.31 MiB       0  [emitted] [dev]         romcal
Entrypoint romcal [big] = romcal.bundle.min.js romcal.bundle.min.js.map
  [2] (webpack)/buildin/module.js 497 bytes {0} [built]
[131] (webpack)/buildin/global.js 472 bytes {0} [built]
[132] ./node_modules/moment/locale sync ^\.\/.*$ 3 KiB {0} [optional] [built]
[134] ./src/index.js + 72 modules 535 KiB {0} [built]
      | ./src/index.js 2.17 KiB [built]
      | ./src/constants/index.js 322 bytes [built]
      | ./src/calendars/index.js 2.28 KiB [built]
      | ./src/locales/index.js 274 bytes [built]
      | ./src/lib/index.js 251 bytes [built]
      | ./src/constants/Cycles.js 90 bytes [built]
      | ./src/constants/LiturgicalColors.js 419 bytes [built]
      | ./src/constants/PsalterWeeks.js 140 bytes [built]
      | ./src/constants/LiturgicalSeasons.js 283 bytes [built]
      | ./src/constants/Titles.js 255 bytes [built]
      | ./src/constants/Types.js 5.35 KiB [built]
      | ./src/calendars/argentina.js 4.95 KiB [built]
      | ./src/calendars/australia.js 1.77 KiB [built]
      | ./src/calendars/austria.js 557 bytes [built]
      | ./src/calendars/belgium.js 2.71 KiB [built]
      |     + 58 hidden modules
    + 131 hidden modules

WARNING in asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
This can impact web performance.
Assets: 
  romcal.bundle.min.js (677 KiB)

WARNING in entrypoint size limit: The following entrypoint(s) combined asset size exceeds the recommended limit (244 KiB). This can impact web performance.
Entrypoints:
  romcal (677 KiB)
      romcal.bundle.min.js


WARNING in webpack performance recommendations: 
You can limit the size of your bundles by using import() or require.ensure to lazy load some parts of your application.
For more info visit https://webpack.js.org/guides/code-splitting/
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.9 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.9: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

added 704 packages from 808 contributors and audited 9397 packages in 53.289s
found 129 vulnerabilities (33 moderate, 96 high)
  run `npm audit fix` to fix them, or `npm audit` for details
```

**Comment by emagnier** [comment]
`npm audit fix` isn't related to that and I don't recommend to use it here.
Maybe it's something related to the path, npm is unable to find the binaries in ./node_modules/.bin

Can you try to change all `npm` to `npx` in the script section of your packages.json? 
```
  "scripts": {
    "prepack": "npx run build && npx test",
    "install": "npx run build",
    "build": "npx run build:babel && npx run build:webpack",
    "build:babel": "babel src --out-dir dist",
    "build:webpack": "webpack-cli --config webpack.config.js",
    "test": "mocha 'test/*.js'"
  },
```

**Comment by tukusejssirs** [comment]
Here’s the [`npm audit`](https://gist.github.com/tukusejssirs/fbb98bf93c40104b2a53b65701ecea61) output; it might be helpful. :smiley:

**Comment by emagnier** [comment]
Ok I didn't read well your last comments. It's working with a `npm install`. However I think I should do the `npx` fix to avoid potential path issue that could happen for some users.

**Comment by tukusejssirs** [comment]
Okay, I will try using `npx`

Anyway, could you tell me what’s the problem here? Why should `npx` fix the issues that `npm audit` raises?

And the other thing: the issues that `npm audit` raises aren’t (security) patches? Should we not apply them during build?

---

PS—I have tidied up my comments as there were posted multiple times (I have quite slow Internet connection, so GitHub displayed errors that it could not post the messages, but in the end it did send all of them).

**Comment by tukusejssirs** [comment]
1. I’ve just replaced `npm` to `npx` as you have suggested (see below the modified `package.json` file content).
2. Then, I thought you might have wanted me to run `npx install`, but it raised `install: missing file operand` error.
3. Finally, I run `npm install`, but it hangs (see the output below).

<details>
   <summary>
      <code>package.json</code>
   </summary>

```json
{
  "name": "romcal",
  "version": "1.3.0-beta3",
  "description": "Utility library that outputs the Liturgical Calendar used by the Roman Rite (Western Church)",
  "main": "dist/index.js",
  "module": "src/index.js",
  "engines": {
    "node": ">=6.0.0"
  },
  "scripts": {
    "prepack": "npx run build && npx test",
    "install": "npx run build",
    "build": "npx run build:babel && npx run build:webpack",
    "build:babel": "babel src --out-dir dist",
    "build:webpack": "webpack-cli --config webpack.config.js",
    "test": "mocha 'test/*.js'"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/pejulian/romcal.git"
  },
  "keywords": [
    "roman",
    "catholic",
    "calendar",
    "latin",
    "western",
    "church",
    "christian",
    "seasons",
    "liturgical"
  ],
  "author": {
    "name": "Julian Pereira",
    "email": "bleushade@gmail.com",
    "url": "https://github.com/pejulian"
  },
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/pejulian/romcal/issues"
  },
  "homepage": "https://github.com/pejulian/romcal#README",
  "dependencies": {
    "lodash": "^4.17.15",
    "moment": "^2.24.0",
    "moment-range": "^4.0.2",
    "moment-recur": "^1.0.7"
  },
  "devDependencies": {
    "@babel/cli": "^7.7.5",
    "@babel/core": "^7.7.5",
    "@babel/node": "^7.7.4",
    "@babel/preset-env": "^7.7.6",
    "@babel/register": "^7.7.4",
    "babel-eslint": "^10.0.3",
    "babel-loader": "^8.0.6",
    "eslint": "^6.7.2",
    "eslint-plugin-node": "^10.0.0",
    "mocha": "^6.2.2",
    "should": "^13.2.3",
    "source-map-loader": "^0.2.4",
    "webpack": "^4.41.2",
    "webpack-cli": "^3.3.10"
  }
}
```
</details>

<details>
   <summary>
      <code>npm install</code>
   </summary>

```bash
$ npm install

> core-js@3.4.8 postinstall /home/ts/git/c10n/romcal_test/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"

Thank you for using core-js ( https://github.com/zloirock/core-js ) for polyfilling JavaScript standard library!

The project needs your help! Please consider supporting of core-js on Open Collective or Patreon: 
> https://opencollective.com/core-js 
> https://www.patreon.com/zloirock 

Also, the author of core-js ( https://github.com/zloirock ) is looking for a good job -)


> romcal@1.3.0-beta3 install /home/ts/git/c10n/romcal_test
> npx run build

npx: installed 5 in 2.646s
Watching /home/ts/git/c10n/romcal_test and all sub-directories not excluded by your .gitignore. Will not monitor dotfiles.
Found & ignored ./node_modules ; is listed in .gitignore

Starting: build
internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module '/home/ts/git/c10n/romcal_test/build'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Function.Module.runMain (internal/modules/cjs/loader.js:831:12)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:622:3)
# And here it stops and it does not continue nor exit.
```
</details>

**Comment by emagnier** [comment]
Before editing the `npm` to `npx`, is your build running with `npm install`?
The `npx` was just a suggestion if your installation wasn't working, but I found that it's not necessary to use them here (and it's not really made for that).

The `npm audit` have nothing related to your issue (as I wrote already). It's a good indication, but I don't recommend to use the `npm audit fix` for the same reasons described in this article: https://www.voitanos.io/blog/don-t-be-alarmed-by-vulnerabilities-after-running-npm-install
For information, they mainly come from webpack and babel-cli, which are dev dependencies, so not installed in production.

**Comment by tukusejssirs** [comment]
> Before editing the `npm` to `npx`, is your build running with `npm install`?

Well, the output is just I have already posted in my original problem. Though I can open `index.html` and see the output in the console, in that console there a `Security Error: Content at file:///home/ts/git/c10n/romcal/index.html may not load or link to moz-extension://bfc6bc5d-a6d9-4778-a75c-15b5ee24a92a/content.css.`.

![Screenshot from 2019-12-15 22-15-38](https://user-images.githubusercontent.com/5111894/70869170-752b3280-1f88-11ea-9acd-a5f583f3e1ff.png)

> The `npm audit` have nothing related to your issue (as I wrote already).

I see what you mean and thanks for the link. But it’d be nice to have those ‘errors/issues’ fixed. :smiley:

**Comment by emagnier** [comment]
Your output doesn't say if your npm install did work or not.

> Before editing the npm to npx, is your build running with npm install?

I just want a yes or no, please reads my comments and answers my questions, so I can try to help you :) It's working every else where (including travis) so for now it's hard to catch what issue you might have.

The CSS issue you have might not be related to romcal. I'm not able to reproduce it on Chrome or Firefox.

> I see what you mean and thanks for the link. But it’d be nice to have those ‘errors/issues’ fixed. 😃

I do all my best Mr. @tukusejssirs ;)

So if your version of romcal is installing and building, let's close this issue.

**Comment by tukusejssirs** [comment]
> Before editing the `npm` to `npx`, is your build running with `npm install`?

Well, when I open `index.html` (either before building, or after building using `npm`, or after building using `npx`), I get the correct output in the console for the Roman Calendar. So it works as expected, but I _thought_ that `npm audit` should not complain at all (but you have already told I was wrong). So this is probably a non-issue.

> I do all my best Mr. @tukusejssirs ;)

I did not want to offend you—all your work is just AWESOME! :smiley:

Anyway, the current `romcal-es6` (33c9f99) shows an additional _source map error_ (see below). And I have used the pre-built `romcal` to get this error.

![Screenshot from 2019-12-16 13-49-25](https://user-images.githubusercontent.com/5111894/70908533-a6007b80-200b-11ea-84e7-8d69726771c8.png)

This link [_Learn More_](https://developer.mozilla.org/en-US/docs/Tools/Debugger/Source_map_errors?utm_source=mozilla&utm_medium=firefox-console-errors&utm_campaign=default) says the following (so this particular issue might be a FF bug):

> **NetworkError when attempting to fetch resource**
>
> A bug in Firefox prevents it from loading source maps for web extensions.
>
> `See Bug 1437937: WebExtensions Doesn't Find Source Maps for details.`
>
> Source-Map-Fehler: TypeError: NetworkError when attempting to fetch resource.
> Ressourcen-Adresse: moz-extension://c7f0f003-4fcf-49fd-8ec0-c49361266581/background.js
> Source-Map-Adresse: background.js.map
>
> The only workaround is to manually change the map URL to a public one (http://localhost:1234/file.map.js) and start a local webserver at this port.

---

And JIT it might help, I use Firefox Nightly on CentOS 8 x86_64.

---

> So if your version of romcal is installing and building, let's close this issue.

Should I open a new issue for those browser console errors?

**Comment by emagnier** [comment]
I'm just trying to explain that `npm audit` have nothing related to the build. And by the way I've fixed the package-lock.js, so `npm` will stop complaining.

The problem you mention in your browser is related to Bootstrap (however I'm not able to reproduce it). This is still a trivial issue so we should not lose too much time on it. And like I recommended already, this file should be as simple as possible, and Bootstrap removed, so we can focus on things that worth it (we should open an issue for that).

I'm closing this issue, since your build is working now 👍
